core/packages/contracts/src/InboundChannel.sol

```diff
@@ -0,0 +1,155 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.19;
+
+import "openzeppelin/utils/cryptography/MerkleProof.sol";
+import "openzeppelin/access/AccessControl.sol";
+import "./ParachainClient.sol";
+import "./IRecipient.sol";
+import "./IVault.sol";
+
+contract InboundChannel is AccessControl {
+    // Nonce for each origin
+    mapping(bytes => uint64) public nonce;
+
+    // Registered message handlers
+    mapping(uint16 => Handler) handlers;
+
+    // Light client message verifier
+    IParachainClient public parachainClient;
+
+    // Relayers are rewarded from this vault
+    IVault public vault;
+
+    // The relayer reward for submitting a message
+    uint256 public reward;
+
+    // The governance contract, which is a proxy for Polkadot governance, administers via this role
+    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+
+    // Relayers must provide enough gas to cover message dispatch plus this buffer
+    uint256 public constant GAS_BUFFER = 24000;
+
+    // Inbound message from BridgeHub parachain
+    struct Message {
+        bytes origin;
+        uint64 nonce;
+        uint16 handler;
+        bytes payload;
+    }
+
+    // Message handler entry
+    struct Handler {
+        // Address of message handler which implements IRecipient
+        address recipient;
+        // Amount of gas to forward to message handler
+        uint32 gasToForward;
+    }
+
+    // The result of message dispatch
+    struct DispatchResult {
+        // Whether the dispatch succeeded
+        bool succeeded;
+        // Various error signifiers
+        string errorReason;
+        uint256 errorPanicCode;
+        bytes errorReturnData;
+    }
+
+    event MessageDispatched(bytes origin, uint64 nonce, DispatchResult result);
+    event HandlerUpdated(uint16 id, Handler handler);
+    event ParachainClientUpdated(address parachainClient);
+    event VaultUpdated(address vault);
+    event RewardUpdated(uint256 reward);
+
+    error InvalidProof();
+    error InvalidNonce();
+    error InvalidHandler();
+    error NotEnoughGas();
+
+    constructor(IParachainClient _parachainClient, IVault _vault, uint256 _reward) {
+        _grantRole(ADMIN_ROLE, msg.sender);
+        parachainClient = _parachainClient;
+        vault = _vault;
+        reward = _reward;
+    }
+
+    function submit(
+        Message calldata message,
+        bytes32[] calldata leafProof,
+        bytes calldata headerProof
+    ) external {
+        bytes32 leafHash = keccak256(abi.encode(message));
+        bytes32 commitment = MerkleProof.processProof(leafProof, leafHash);
+        if (!parachainClient.verifyCommitment(commitment, headerProof)) {
+            revert InvalidProof();
+        }
```

Comment:
comment explaining each line would be helpful here.

What is the commitment in the first line?

What is being verified in the second line? That the commitment is in a parachain header or?

Reply:
One major thing happening here is also that we are verifying the origin here that was provided as input in the message.

This is critical to understand for a reader of the code, since all security flows from trust in that origin and the authentication being done.

Since this is a bit of a base axiom that one needs to accept while auditing/reviewing the code, may be worth to have a comment on it here, and definitely as part of a README/document we should have a clear explanation of this and of why the authentication process can be trusted.

Reply:
Another question: origin here is in practice basically a parachain, right?

What about in general - can it be a specific pallet or even specific account in a smart contract in the VM of a parachain?

What about parachains that have smart contracts where users can control code/messages, like Moonbeam where if they exposed this XCM to users then those users could send messages as if they were the parachain to withdraw other users funds? Is this the case? Will we expect Moonbeam to secure this on their own on their side in this case or what? How do then expect an integration/implementation inside their smart contract VM to look?

Reply:
> Another question: origin here is in practice basically a parachain, right?

Correct

> What about in general - can it be a specific pallet or even specific account in a smart contract in the VM of a parachain?

Nope, must always be a parachain, whose governance must take responsibility for charging fees from sub-consensus systems (pallet, etc) fees for using their dedicated messaging lane.

For example an EVM contract on Acala must pay fees to Acala toplevel governance for using Acala's messaging lane.

As for malicious sub-consensus systems, the bridge is designed to be resilient against them. Our pallets on BridgeHub will filter messages to only allow specific kinds of messages to be sent to Ethereum.